### ISSUE TITLE
k3s agent "only https:// URLs are supported, invalid scheme"

### ISSUE THREAD

**Issue by mcsaucy** [rancher/k3s#1964]
**Version:**

**K3s arguments:**

`[core@node1 ~]$ curl -sfL https://get.k3s.io | K3S_URL=http://myserver:6443 K3S_TOKEN=mynodetoken sh -` (note the `http://` in K3S_URL)

**Describe the bug**
Install script succeeds, but k3s-agent fails to start with `Jun 27 15:05:20 node1 k3s[2748]: time="2020-06-27T15:05:20.405611746Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:40035"`

**To Reproduce**
Run `curl -sfL https://get.k3s.io | K3S_URL=http://myserver:6443 K3S_TOKEN=mynodetoken sh -`

**Expected behavior**
The install script should bail rather than propagating an invalid parameter

**Actual behavior**
Script exits successfully, but `k3s agent` fails to start with the given parameters.

**Additional context / logs**
Providing an HTTPS `K3S_URL` (such as literally https://myserver:6443) makes it past this check, which suggests we're setting up a local proxy or something to `K3S_URL` matching the same schema. We then error out on *that*, which is somewhat confusing.

**Comment by ShylajaDevadiga** [comment]
@mcsaucy Logs do not seem to reflect the changes in install script while passing http:// in K3S_URL
Installed k3s using commit id: e3f87891142a38aa4e6194c76741f17f30911ee8 

```
k3s -v
k3s version v1.18.6+k3s-e3f87891 (e3f87891)
```
With http in K3S_URL=http:// logs 
```
Jul 21 18:42:48 ip-172-31-2-104 k3s[11927]: time="2020-07-21T18:42:48.141583586Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:40219"
Jul 21 18:42:50 ip-172-31-2-104 k3s[11927]: time="2020-07-21T18:42:50.141911411Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:40219"
```

**Comment by brandond** [comment]
@ShylajaDevadiga how are you testing?

```
python38) brandond@dev01:~$ curl -sfL https://get.k3s.io | K3S_TOKEN=secret K3S_URL=http://test:6443 sh -
[ERROR]  Only https:// URLs are supported for K3S_URL (have http://test:6443)
```

**Comment by ShylajaDevadiga** [comment]
@brandond I am trying to join an agent to a master.  
```
curl -sfL https://get.k3s.io | INSTALL_K3S_COMMIT=04f57e5e1da489278d039478f095deb74bcc9b54 K3S_URL=http://IP:6443 K3S_TOKEN=secret sh -
[ERROR]  Only https:// URLs are supported for K3S_URL (have http://IP:6443)
```
When i actually join an agent
```
curl -sfL https://get.k3s.io | INSTALL_K3S_COMMIT=04f57e5e1da489278d039478f095deb74bcc9b54 K3S_URL=http://3.136.26.126:6443 K3S_TOKEN=secret sh -s - agent
[INFO]  Using commit 04f57e5e1da489278d039478f095deb74bcc9b54 as release
[INFO]  Downloading hash https://storage.googleapis.com/k3s-ci-builds/k3s-04f57e5e1da489278d039478f095deb74bcc9b54.sha256sum
[INFO]  Skipping binary downloaded, installed k3s matches hash
```
Logs:
```
Jul 23 19:18:31 ip-172-31-9-149 k3s[3983]: time="2020-07-23T19:18:31.845350270Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:32841"
Jul 23 19:18:33 ip-172-31-9-149 k3s[3983]: time="2020-07-23T19:18:33.845610443Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:32841"
```

**Comment by brandond** [comment]
I didn't look closely at the script. The current code only checks K3S_URL if there are no arguments ie neither `server` nor `agent` is explicitly requested. The URL should be checked either way.

tl;dr this is checked and fails immediately:
`curl -sfL https://get.k3s.io | K3S_TOKEN=secret K3S_URL=http://test:6443 sh -`
this is not checked and does not fail until trying to start the service:
`curl -sfL https://get.k3s.io | K3S_TOKEN=secret K3S_URL=http://test:6443 sh - agent`

**Comment by mcsaucy** [comment]
I'll move that check out of that condition.

**Comment by brandond** [comment]
Yeah, either move it out and up, or maybe put it in a function that's called from the right places?

**Comment by ShylajaDevadiga** [comment]
Validated the fix using k3s version v1.18.6+k3s-169ee639
```
Master
curl -sfL http://get.k3s.io |INSTALL_K3S_COMMIT=169ee63907a51004ff000af1238e3483ce0c5fd8 sh -

Agent
curl -sfL http://get.k3s.io INSTALL_K3S_COMMIT=169ee63907a51004ff000af1238e3483ce0c5fd8 |  K3S_URL=http://IP:6443 K3S_TOKEN=TOKEN sh -
[ERROR]  Only https:// URLs are supported for K3S_URL (have http://IP:6443)
```

**Comment by HairlessVillager** [comment]
Met the same problem using `rancher/k3s` image:

```
docker run -d --env-file k3s-agent-env.txt rancher/k3s agent
```

The `k3s-agent-env.txt` looks like:
```
K3S_URL=http://myserver:6443
K3S_TOKEN=xxx
```

Log:
```
time="2025-04-29T15:50:21Z" level=warning msg="cgroup2 is not mounted. cgroup2 evacuation is discarded."
time="2025-04-29T15:50:21Z" level=info msg="Starting k3s agent v1.29.15+k3s1 (35a47239)"
time="2025-04-29T15:50:21Z" level=info msg="Updated load balancer k3s-agent-load-balancer default server: myserver:6443"
time="2025-04-29T15:50:21Z" level=info msg="Running load balancer k3s-agent-load-balancer 127.0.0.1:6444 -> [] [default: myserver:6443]"
time="2025-04-29T15:50:21Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:6444"
time="2025-04-29T15:50:23Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:6444"
time="2025-04-29T15:50:25Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:6444"
time="2025-04-29T15:50:27Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:6444"
time="2025-04-29T15:50:29Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:6444"
time="2025-04-29T15:50:31Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:6444"
time="2025-04-29T15:50:33Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:6444"
time="2025-04-29T15:50:35Z" level=error msg="only https:// URLs are supported, invalid scheme: http://127.0.0.1:6444"'
```

Solved by just replacing "http://myserver:6443" with "https://myserver:6443"

**Comment by brandond** [comment]
yes... server URL is never https.
